Iter18

abezemskiy opened the pull request.

abezemskiy committed:
Fix method of building errors in test function

abezemskiy committed:
Add function for cleaning database

abezemskiy committed:
Add benchmark in handlers package

abezemskiy committed:
Add end-to-end testing for getting heap profile

abezemskiy committed:
Use prepare sql queries

abezemskiy committed:
Merge branch 'main' into iter16

abezemskiy committed:
Change go version in actions

abezemskiy committed:
Merge branch 'iter16' of github.com:AntonBezemskiy/go-musthave-metrics into iter16

abezemskiy committed:
Change go version in actions to 1.23.1

abezemskiy committed:
Format project with using goimports

abezemskiy committed:
Add test functions

abezemskiy committed:
Add comments for documentation

abezemskiy committed:
Add example for GetMetricJSON handler

abezemskiy committed:
Fix end-to-end test for getting heap profile

abezemskiy committed:
Fix end-to-end test for getting heap profile

abezemskiy committed:
Change go version

abezemskiy committed:
Change go version

abezemskiy committed:
Add autotests for 16-18 increments

abezemskiy committed:
Change storage in bd to in memory in benchmark

Perederey commented on internal/server/pg/store.go:
```diff
@@ -6,9 +6,11 @@ import (
+// Store реализует интерфейс store.Store и позволяет взаимодействовать с СУБД PostgreSQL.
+// Так же Store реализует интерфейс repositories.ServerRepo, для возможности использования структуры в качестве хранилища метрик.
 type Store struct {
 	// Поле conn содержит объект соединения с СУБД
```
Лучше `sql.DB` вместо `*sql.DB`

abezemskiy replied:
Можно пожалуйста объяснить, почему?
В пользу `*sql.DB` есть пара аргументов:
1. `sql.DB` содержит поле `sync.Mutex`, которое не может быть скопировано безопасно (что описано в [стайлгайде](https://google.github.io/styleguide/go/decisions#receiver-type)).
2. `sql.DB` создает пул соединений и копирование этой структуры нежелательно.

Perederey commented on internal/agent/handlers/handlers.go:
```diff
@@ -14,15 +14,16 @@
 	"syscall"
```
Вынеси логику работы с метриками в отдельный пакет, чтобы улучшить разделение ответственности и упростить тестирование

abezemskiy replied:
Разделил логику в ecb57c3cd6956bad825577a969063025fcd205e1

Perederey commented on internal/repositories/repositories.go:
```diff
@@ -5,23 +5,25 @@ import (
 	"fmt"
 )
 
+// Интерфесы хранилища метрик.
```
Раздели интерфейсы на более мелкие, специализированные интерфейсы для лучшей гибкости

abezemskiy replied:
Разделил интерфейсы в a90aa8d3aa34fcf8afbeaa0cf205454b59ec5a02

Perederey commented on internal/server/saver/saver.go:
```diff
@@ -21,47 +21,58 @@ var (
 
 // end Global variable -------------------------------------------------
 
+// WriterInterface - интерфейс записи метрик.
```
Лучше `MetricWriter` для большей конкретики и соответствия Go-стилю

abezemskiy replied:
Исправил в 21abc6caf6f755a266053ad1507bbe9aa37f4a4b

Perederey commented on internal/server/handlers/example_test.go:
```diff
@@ -0,0 +1,104 @@
+	"github.com/AntonBezemskiy/go-musthave-metrics/internal/server/storage"
+)
+
+func ExampleGetMetricJSON() {
```
Тут отлично

Perederey reviewed (approved):
Привет! Ты проделал отличную работу, но есть несколько моментов, которые стоит улучшить:
- Обрати внимание на именование интерфейсов и методов, чтобы они лучше отражали их функциональность
- Разделяй логику обработки HTTP-запросов и бизнес-логику для улучшения архитектуры

Ты хорошо поработал над комментариями и использованием интерфейсов. Продолжай в том же духе

abezemskiy committed:
Rename interfaces in internal/server/saver

abezemskiy committed:
Split interfaces in internal/repositories

abezemskiy committed:
Refactor agent package structure

- Separate metrics handling into dedicated packages
- Move server push logic to its own package
